// when/lib/Scheduler.js
System.registerModule("../../../../node_modules/when/lib/Scheduler.js", [], function() {
  "use strict";
  var __moduleName = "../../../../node_modules/when/lib/Scheduler.js";
  (function(define) {
    'use strict';
    define(function() {
      function Scheduler(async) {
        this._async = async;
        this._running = false;
        this._queue = this;
        this._queueLen = 0;
        this._afterQueue = {};
        this._afterQueueLen = 0;
        var self = this;
        this.drain = function() {
          self._drain();
        };
      }
      Scheduler.prototype.enqueue = function(task) {
        this._queue[this._queueLen++] = task;
        this.run();
      };
      Scheduler.prototype.afterQueue = function(task) {
        this._afterQueue[this._afterQueueLen++] = task;
        this.run();
      };
      Scheduler.prototype.run = function() {
        if (!this._running) {
          this._running = true;
          this._async(this.drain);
        }
      };
      Scheduler.prototype._drain = function() {
        var i = 0;
        for (; i < this._queueLen; ++i) {
          this._queue[i].run();
          this._queue[i] = void 0;
        }
        this._queueLen = 0;
        this._running = false;
        for (i = 0; i < this._afterQueueLen; ++i) {
          this._afterQueue[i].run();
          this._afterQueue[i] = void 0;
        }
        this._afterQueueLen = 0;
      };
      return Scheduler;
    });
  }(typeof define === 'function' && define.amd ? define : function(factory) {
    module.exports = factory();
  }));
  return {};
});
System.get("../../../../node_modules/when/lib/Scheduler.js" + '');
