// when/lib/state.js
System.registerModule("../../../../node_modules/when/lib/state.js", [], function() {
  "use strict";
  var __moduleName = "../../../../node_modules/when/lib/state.js";
  (function(define) {
    'use strict';
    define(function() {
      return {
        pending: toPendingState,
        fulfilled: toFulfilledState,
        rejected: toRejectedState,
        inspect: inspect
      };
      function toPendingState() {
        return {state: 'pending'};
      }
      function toRejectedState(e) {
        return {
          state: 'rejected',
          reason: e
        };
      }
      function toFulfilledState(x) {
        return {
          state: 'fulfilled',
          value: x
        };
      }
      function inspect(handler) {
        var state = handler.state();
        return state === 0 ? toPendingState() : state > 0 ? toFulfilledState(handler.value) : toRejectedState(handler.value);
      }
    });
  }(typeof define === 'function' && define.amd ? define : function(factory) {
    module.exports = factory();
  }));
  return {};
});
System.get("../../../../node_modules/when/lib/state.js" + '');
