// when/monitor/error.js
System.registerModule("../../../../node_modules/when/monitor/error.js", [], function() {
  "use strict";
  var __moduleName = "../../../../node_modules/when/monitor/error.js";
  (function(define) {
    'use strict';
    define(function() {
      var parse,
          captureStack,
          format;
      if (Error.captureStackTrace) {
        parse = function(e) {
          return e && e.stack && e.stack.split('\n');
        };
        format = formatAsString;
        captureStack = Error.captureStackTrace;
      } else {
        parse = function(e) {
          var stack = e && e.stack && e.stack.split('\n');
          if (stack && e.message) {
            stack.unshift(e.message);
          }
          return stack;
        };
        (function() {
          var e = new Error();
          if (typeof e.stack !== 'string') {
            format = formatAsString;
            captureStack = captureSpiderMonkeyStack;
          } else {
            format = formatAsErrorWithStack;
            captureStack = useStackDirectly;
          }
        }());
      }
      function captureSpiderMonkeyStack(host) {
        try {
          throw new Error();
        } catch (err) {
          host.stack = err.stack;
        }
      }
      function useStackDirectly(host) {
        host.stack = new Error().stack;
      }
      function formatAsString(longTrace) {
        return join(longTrace);
      }
      function formatAsErrorWithStack(longTrace) {
        var e = new Error();
        e.stack = formatAsString(longTrace);
        return e;
      }
      function join(a) {
        var sep = false;
        var s = '';
        for (var i = 0; i < a.length; ++i) {
          if (sep) {
            s += '\n' + a[i];
          } else {
            s += a[i];
            sep = true;
          }
        }
        return s;
      }
      return {
        parse: parse,
        format: format,
        captureStack: captureStack
      };
    });
  }(typeof define === 'function' && define.amd ? define : function(factory) {
    module.exports = factory();
  }));
  return {};
});
System.get("../../../../node_modules/when/monitor/error.js" + '');
