// zone.js/lib/jasmine/patch.js
System.registerModule("../../../../../node_modules/zone.js/lib/jasmine/patch.js", [], function() {
  "use strict";
  var __moduleName = "../../../../../node_modules/zone.js/lib/jasmine/patch.js";
  function apply() {
    if (!global.zone) {
      throw new Error('zone.js does not seem to be installed');
    }
    if (!global.zone.isRootZone()) {
      throw new Error('The jasmine patch should be called from the root zone');
    }
    var jasmineZone = global.zone;
    var originalIt = global.it;
    var originalFit = global.fit;
    var patchJasmineDone = function(jasmineDone) {
      var done = jasmineZone.bind(jasmineDone);
      if (typeof jasmineDone.fail === 'function') {
        done.fail = jasmineZone.bind(jasmineDone.fail);
      }
      return done;
    };
    global.it = function zoneResettingIt(description, specFn, timeOut) {
      if (specFn.length) {
        originalIt(description, function zoneResettingSpecFn(done) {
          specFn(patchJasmineDone(done));
        }, timeOut);
      } else {
        originalIt(description, specFn, timeOut);
      }
    };
    global.fit = function zoneResettingFit(description, specFn, timeOut) {
      if (specFn.length) {
        originalFit(description, function zoneResettingSpecFn(done) {
          specFn(patchJasmineDone(done));
        }, timeOut);
      } else {
        originalFit(description, specFn, timeOut);
      }
    };
    beforeEach(function() {
      expect(global.zone.isRootZone()).toBe(true);
    });
  }
  if (global.jasmine) {
    module.exports = {apply: apply};
  } else {
    module.exports = {apply: function() {}};
  }
  return {};
});
System.get("../../../../../node_modules/zone.js/lib/jasmine/patch.js" + '');
