// foxy/lib/config.js
System.registerModule("../../../../node_modules/foxy/lib/config.js", [], function() {
  "use strict";
  var __moduleName = "../../../../node_modules/foxy/lib/config.js";
  var url = require("url");
  var utils = require("./utils");
  var errors = require("./errors");
  var merge = require("lodash.merge");
  var defaults = {
    errHandler: errors,
    whitelist: [],
    blacklist: [],
    cookies: {stripDomain: true},
    reqHeaders: function(config) {
      return {
        "host": config.urlObj.host,
        "accept-encoding": "identity",
        "agent": false
      };
    },
    staticFiles: {},
    middleware: [utils.handleIe],
    proxyOptions: {
      xfwd: false,
      headers: {},
      target: "",
      secure: false,
      ws: true
    },
    proxyRes: [utils.checkCookies, utils.removeExcludedHeaders, utils.handleRedirect],
    excludedHeaders: ["content-length", "content-encoding"]
  };
  module.exports = function(target, userConfig) {
    var config = merge({}, defaults, userConfig, function(a, b) {
      if (Array.isArray(a)) {
        return a.concat(b);
      }
    });
    var urlObj = url.parse(target);
    var _target = urlObj.protocol + "//" + urlObj.hostname;
    config.urlObj = urlObj;
    config.hostHeader = utils.getProxyHost(urlObj);
    if (urlObj.port) {
      _target = _target + ":" + urlObj.port;
    }
    config.target = _target;
    return config;
  };
  return {};
});
System.get("../../../../node_modules/foxy/lib/config.js" + '');
