// angular2/src/platform/browser/ruler.js
System.registerModule("../../../../../../node_modules/angular2/src/platform/browser/ruler.js", [], function() {
  "use strict";
  var __moduleName = "../../../../../../node_modules/angular2/src/platform/browser/ruler.js";
  var async_1 = require('angular2/src/facade/async');
  var Rectangle = (function() {
    function Rectangle(left, top, width, height) {
      this.left = left;
      this.right = left + width;
      this.top = top;
      this.bottom = top + height;
      this.height = height;
      this.width = width;
    }
    return Rectangle;
  })();
  exports.Rectangle = Rectangle;
  var Ruler = (function() {
    function Ruler(domAdapter) {
      this.domAdapter = domAdapter;
    }
    Ruler.prototype.measure = function(el) {
      var clntRect = this.domAdapter.getBoundingClientRect(el.nativeElement);
      return async_1.PromiseWrapper.resolve(new Rectangle(clntRect.left, clntRect.top, clntRect.width, clntRect.height));
    };
    return Ruler;
  })();
  exports.Ruler = Ruler;
  return {};
});
System.get("../../../../../../node_modules/angular2/src/platform/browser/ruler.js" + '');
