// browser-sync-ui/lib/plugins/history/history.plugin.js
System.registerModule("../../../../../../node_modules/browser-sync-ui/lib/plugins/history/history.plugin.js", [], function() {
  "use strict";
  var __moduleName = "../../../../../../node_modules/browser-sync-ui/lib/plugins/history/history.plugin.js";
  var historyPlugin = require("./history");
  var PLUGIN_NAME = "History";
  module.exports = {
    "plugin": function(ui, bs) {
      ui.history = historyPlugin.init(ui, bs);
    },
    "hooks": {
      "markup": fileContent("history.html"),
      "client:js": fileContent("/history.client.js"),
      "templates": [getPath("/history.directive.html")],
      "page": {
        path: "/history",
        title: PLUGIN_NAME,
        template: "history.html",
        controller: PLUGIN_NAME + "Controller",
        order: 3,
        icon: "list2"
      }
    },
    "plugin:name": PLUGIN_NAME
  };
  function getPath(filepath) {
    return require("path").join(__dirname, filepath);
  }
  function fileContent(filepath) {
    return require("fs").readFileSync(getPath(filepath), "utf-8");
  }
  return {};
});
System.get("../../../../../../node_modules/browser-sync-ui/lib/plugins/history/history.plugin.js" + '');
