// tough-cookie/lib/store.js
System.registerModule("../../../../../../node_modules/fsevents/node_modules/tough-cookie/lib/store.js", [], function() {
  "use strict";
  var __moduleName = "../../../../../../node_modules/fsevents/node_modules/tough-cookie/lib/store.js";
  function Store() {}
  exports.Store = Store;
  Store.prototype.synchronous = false;
  Store.prototype.findCookie = function(domain, path, key, cb) {
    throw new Error('findCookie is not implemented');
  };
  Store.prototype.findCookies = function(domain, path, cb) {
    throw new Error('findCookies is not implemented');
  };
  Store.prototype.putCookie = function(cookie, cb) {
    throw new Error('putCookie is not implemented');
  };
  Store.prototype.updateCookie = function(oldCookie, newCookie, cb) {
    throw new Error('updateCookie is not implemented');
  };
  Store.prototype.removeCookie = function(domain, path, key, cb) {
    throw new Error('removeCookie is not implemented');
  };
  Store.prototype.removeCookies = function(domain, path, cb) {
    throw new Error('removeCookies is not implemented');
  };
  Store.prototype.getAllCookies = function(cb) {
    throw new Error('getAllCookies is not implemented (therefore jar cannot be serialized)');
  };
  return {};
});
System.get("../../../../../../node_modules/fsevents/node_modules/tough-cookie/lib/store.js" + '');
