// browser-sync/lib/public/public-utils.js
System.registerModule("../../../../../node_modules/browser-sync/lib/public/public-utils.js", [], function() {
  "use strict";
  var __moduleName = "../../../../../node_modules/browser-sync/lib/public/public-utils.js";
  var _ = require("lodash");
  module.exports = {
    emitChangeEvent: function emitChangeEvent(emitter, path, log) {
      emitter.emit("file:changed", {
        path: path,
        log: log,
        namespace: "core"
      });
    },
    emitBrowserReload: function emitChangeEvent(emitter) {
      emitter.emit("browser:reload");
    },
    emitStreamChangedEvent: function(emitter, changed) {
      emitter.emit("stream:changed", {changed: changed});
    },
    isStreamArg: function(name, args) {
      if (name === "stream") {
        return true;
      }
      if (name !== "reload") {
        return false;
      }
      var firstArg = args[0];
      if (_.isObject(firstArg)) {
        if (!Array.isArray(firstArg) && Object.keys(firstArg).length) {
          return firstArg.stream === true;
        }
      }
      return false;
    }
  };
  return {};
});
System.get("../../../../../node_modules/browser-sync/lib/public/public-utils.js" + '');
