// browser-sync/lib/default-config.js
System.registerModule("../../../../node_modules/browser-sync/lib/default-config.js", [], function() {
  "use strict";
  var __moduleName = "../../../../node_modules/browser-sync/lib/default-config.js";
  module.exports = {
    ui: {
      port: 3001,
      weinre: {port: 8080}
    },
    files: false,
    watchOptions: {},
    server: false,
    proxy: false,
    port: 3000,
    middleware: false,
    serveStatic: [],
    ghostMode: {
      clicks: true,
      scroll: true,
      forms: {
        submit: true,
        inputs: true,
        toggles: true
      }
    },
    logLevel: "info",
    logPrefix: "BS",
    logConnections: false,
    logFileChanges: true,
    logSnippet: true,
    snippetOptions: {
      async: true,
      whitelist: [],
      blacklist: [],
      rule: {
        match: /<body[^>]*>/i,
        fn: function(snippet, match) {
          return match + snippet;
        }
      }
    },
    rewriteRules: false,
    open: "local",
    browser: "default",
    xip: false,
    hostnameSuffix: false,
    reloadOnRestart: false,
    notify: true,
    scrollProportionally: true,
    scrollThrottle: 0,
    scrollRestoreTechnique: "window.name",
    scrollElements: [],
    scrollElementMapping: [],
    reloadDelay: 0,
    reloadDebounce: 0,
    plugins: [],
    injectChanges: true,
    startPath: null,
    minify: true,
    host: null,
    codeSync: true,
    timestamps: true,
    clientEvents: ["scroll", "scroll:element", "input:text", "input:toggles", "form:submit", "form:reset", "click"],
    socket: {
      socketIoOptions: {log: false},
      socketIoClientConfig: {reconnectionAttempts: 50},
      path: "/browser-sync/socket.io",
      clientPath: "/browser-sync",
      namespace: "/browser-sync",
      clients: {heartbeatTimeout: 5000}
    },
    tagNames: {
      "less": "link",
      "scss": "link",
      "css": "link",
      "jpg": "img",
      "jpeg": "img",
      "png": "img",
      "svg": "img",
      "gif": "img",
      "js": "script"
    },
    injectFileTypes: ["css", "png", "jpg", "jpeg", "svg", "gif", "webp"],
    excludedFileTypes: ["js", "css", "pdf", "map", "svg", "ico", "woff", "json", "eot", "ttf", "png", "jpg", "jpeg", "webp", "gif", "mp4", "mp3", "3gp", "ogg", "ogv", "webm", "m4a", "flv", "wmv", "avi", "swf", "scss"]
  };
  return {};
});
System.get("../../../../node_modules/browser-sync/lib/default-config.js" + '');
