// browser-sync/lib/file-utils.js
System.registerModule("../../../../node_modules/browser-sync/lib/file-utils.js", [], function() {
  "use strict";
  var __moduleName = "../../../../node_modules/browser-sync/lib/file-utils.js";
  var _ = require("lodash");
  var fileUtils = {
    changedFile: function(bs, data) {
      if (_.isUndefined(data.event)) {
        data.event = "change";
      }
      if (data.event === "change") {
        if (!bs.paused && data.namespace === "core") {
          bs.events.emit("file:reload", fileUtils.getFileInfo(data, bs.options));
        }
      }
    },
    getFileInfo: function(data, options) {
      data.ext = require("path").extname(data.path).slice(1);
      data.basename = require("path").basename(data.path);
      var obj = {
        ext: data.ext,
        path: data.path,
        basename: data.basename,
        type: "inject"
      };
      if (!_.contains(options.get("injectFileTypes").toJS(), obj.ext)) {
        obj.url = obj.path;
        obj.type = "reload";
      }
      obj.path = data.path;
      obj.log = data.log;
      return obj;
    }
  };
  module.exports = fileUtils;
  return {};
});
System.get("../../../../node_modules/browser-sync/lib/file-utils.js" + '');
