// systemjs/lib/proto.js
System.registerModule("../../../../node_modules/systemjs/lib/proto.js", [], function() {
  "use strict";
  var __moduleName = "../../../../node_modules/systemjs/lib/proto.js";
  function SystemJSLoader() {
    SystemLoader.call(this);
    systemJSConstructor.call(this);
  }
  function SystemProto() {}
  ;
  SystemProto.prototype = SystemLoader.prototype;
  SystemJSLoader.prototype = new SystemProto();
  SystemJSLoader.prototype.constructor = SystemJSLoader;
  var systemJSConstructor;
  function hook(name, hook) {
    SystemJSLoader.prototype[name] = hook(SystemJSLoader.prototype[name] || function() {});
  }
  function hookConstructor(hook) {
    systemJSConstructor = hook(systemJSConstructor || function() {});
  }
  function dedupe(deps) {
    var newDeps = [];
    for (var i = 0,
        l = deps.length; i < l; i++)
      if (indexOf.call(newDeps, deps[i]) == -1)
        newDeps.push(deps[i]);
    return newDeps;
  }
  function group(deps) {
    var names = [];
    var indices = [];
    for (var i = 0,
        l = deps.length; i < l; i++) {
      var index = indexOf.call(names, deps[i]);
      if (index === -1) {
        names.push(deps[i]);
        indices.push([i]);
      } else {
        indices[index].push(i);
      }
    }
    return {
      names: names,
      indices: indices
    };
  }
  var getOwnPropertyDescriptor = true;
  try {
    Object.getOwnPropertyDescriptor({a: 0}, 'a');
  } catch (e) {
    getOwnPropertyDescriptor = false;
  }
  function getESModule(exports) {
    var esModule = {};
    if (typeof exports == 'object' || typeof exports == 'function') {
      if (getOwnPropertyDescriptor) {
        var d;
        for (var p in exports)
          if (d = Object.getOwnPropertyDescriptor(exports, p))
            defineProperty(esModule, p, d);
      } else {
        var hasOwnProperty = exports && exports.hasOwnProperty;
        for (var p in exports) {
          if (!hasOwnProperty || exports.hasOwnProperty(p))
            esModule[p] = exports[p];
        }
      }
    }
    esModule['default'] = exports;
    defineProperty(esModule, '__useDefault', {value: true});
    return esModule;
  }
  function extend(a, b, prepend) {
    for (var p in b) {
      if (!prepend || !(p in a))
        a[p] = b[p];
    }
    return a;
  }
  var packageProperties = ['main', 'format', 'defaultExtension', 'modules', 'map', 'basePath', 'depCache'];
  function extendMeta(a, b, prepend) {
    for (var p in b) {
      var val = b[p];
      if (!(p in a))
        a[p] = val;
      else if (val instanceof Array && a[p] instanceof Array)
        a[p] = [].concat(prepend ? val : a[p]).concat(prepend ? a[p] : val);
      else if (typeof val == 'object' && typeof a[p] == 'object')
        a[p] = extend(extend({}, a[p]), val, prepend);
      else if (!prepend)
        a[p] = val;
    }
  }
  function warn(msg) {
    if (this.warnings && typeof console != 'undefined' && console.warn)
      console.warn(msg);
  }
  return {};
});
System.get("../../../../node_modules/systemjs/lib/proto.js" + '');
