// connect/lib/middleware/limit.js
System.registerModule("../../../../../../../node_modules/express/node_modules/connect/lib/middleware/limit.js", [], function() {
  "use strict";
  var __moduleName = "../../../../../../../node_modules/express/node_modules/connect/lib/middleware/limit.js";
  module.exports = function limit(bytes) {
    if ('string' == typeof bytes)
      bytes = parse(bytes);
    if ('number' != typeof bytes)
      throw new Error('limit() bytes required');
    return function limit(req, res, next) {
      var received = 0,
          len = req.headers['content-length'] ? parseInt(req.headers['content-length'], 10) : null;
      function deny() {
        req.destroy();
      }
      if (req._limit)
        return next();
      req._limit = true;
      if (len && len > bytes) {
        res.statusCode = 413;
        res.end('Request Entity Too Large');
        return;
      }
      req.on('data', function(chunk) {
        received += chunk.length;
        if (received > bytes)
          deny();
      });
      next();
    };
  };
  function parse(size) {
    var parts = size.match(/^(\d+(?:\.\d+)?) *(kb|mb|gb)$/),
        n = parseFloat(parts[1]),
        type = parts[2];
    var map = {
      kb: 1024,
      mb: 1024 * 1024,
      gb: 1024 * 1024 * 1024
    };
    return map[type] * n;
  }
  return {};
});
System.get("../../../../../../../node_modules/express/node_modules/connect/lib/middleware/limit.js" + '');
