// bs-recipes/recipes/webpack.react-transform-hmr/webpack.config.js
System.registerModule("../../../../../node_modules/bs-recipes/recipes/webpack.react-transform-hmr/webpack.config.js", [], function() {
  "use strict";
  var __moduleName = "../../../../../node_modules/bs-recipes/recipes/webpack.react-transform-hmr/webpack.config.js";
  var webpack = require('webpack');
  var path = require('path');
  module.exports = {
    debug: true,
    devtool: '#eval-source-map',
    context: path.join(__dirname, 'app', 'js'),
    entry: ['webpack/hot/dev-server', 'webpack-hot-middleware/client', './main'],
    output: {
      path: path.join(__dirname, 'app', 'js'),
      publicPath: '/js/',
      filename: 'bundle.js'
    },
    plugins: [new webpack.optimize.OccurenceOrderPlugin(), new webpack.HotModuleReplacementPlugin(), new webpack.NoErrorsPlugin()],
    module: {loaders: [{
        test: /\.jsx?$/,
        exclude: /node_modules/,
        loaders: ['babel']
      }]}
  };
  return {};
});
System.get("../../../../../node_modules/bs-recipes/recipes/webpack.react-transform-hmr/webpack.config.js" + '');
