// bs-recipes/recipes/grunt.html.injection/Gruntfile.js
System.registerModule("../../../../../node_modules/bs-recipes/recipes/grunt.html.injection/Gruntfile.js", [], function() {
  "use strict";
  var __moduleName = "../../../../../node_modules/bs-recipes/recipes/grunt.html.injection/Gruntfile.js";
  module.exports = function(grunt) {
    grunt.initConfig({
      watch: {
        files: 'app/scss/**/*.scss',
        tasks: ['bsReload:css']
      },
      sass: {dev: {files: {'app/css/main.css': 'app/scss/main.scss'}}},
      browserSync: {dev: {options: {
            watchTask: true,
            server: './app',
            plugins: [{
              module: "bs-html-injector",
              options: {files: "./app/*.html"}
            }]
          }}},
      bsReload: {css: "main.css"}
    });
    grunt.loadNpmTasks('grunt-contrib-sass');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-browser-sync');
    grunt.registerTask('default', ['browserSync', 'watch']);
  };
  return {};
});
System.get("../../../../../node_modules/bs-recipes/recipes/grunt.html.injection/Gruntfile.js" + '');
